File: src/storeFx/fetchOneSong.ts
import * as Fx from '../fx/fx';
import { mergeIntoMap } from '../mergeIntoMap/mergeIntoMap';
import type { Store } from '../store/types';
import { getSong } from '../subsonic';

export function make(id: string) {
  return Fx.bind(Fx.ask<{ store: Store }>(), ({ store }) =>
    Fx.bind(
      Fx.provide(getSong(id), {
        credentials: store.getState().auth.credentials,
      }),
      song => {
        const state = store.getState();
        const byId = mergeIntoMap(state.songs.byId, [song], x => x.id);

        if (byId !== state.songs.byId) {
          store.setState(() => ({ songs: { byId } }));
        }

        return Fx.Ok();
      },
    ),
  );
}

File: src/storeFx/fetchOneSong.ml
open Fx
open MergeIntoMap
open Monad_syntax

let make id =
  let* Deps.{ store } = ask () in
  let* song =
    let state = store |. Zustand.getState in
    Subsonic.GetSong.make id
    |. provide ~deps:{ credentials = state.auth.credentials }
  in
  let state = store |. Zustand.getState in
  let byId = mergeIntoMap state.songs.byId [| song |] (fun x -> x.id) in
  if byId != state.songs.byId then
    store
    |. Zustand.setState (fun prevState -> { prevState with songs = { byId } });
  Ok ()
